### PR TITLE
Add max conn config

### PR DESCRIPTION
**Pull request by Thanhphan1147**
Applicable spec: ISD-163 (internal)

### Overview

Adds a global_maxconn config option that allows modifying the maxconn option in the haproxy global stanza

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Comment by Thanhphan1147** [comment]
Hi @gregory-schiano, could you take a look at this PR and review it when you have the time? Thanks!

File: src/state/config.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    @classmethod
+    def from_charm(cls, charm: ops.CharmBase) -> "CharmConfig":
+        """Create a CharmConfig class from a charm instance.
+
+        Args:
+            charm: The gateway-api-integrator charm.
+
+        Raises:
+            InvalidCharmConfigError: When the chamr's config is invalid.
```
Comment (mthaddon) [inline src/state/config.py]: s/chamr/charm/

**Review by mthaddon — COMMENTED** [review]
One very minor comment, otherwise looks good

File: src/state/config.py (hunk 1 above)
Comment (Thanhphan1147) [reply src/state/config.py]: Thanks, this was fixed in future PRs but I haven't back ported it to this one. I just fixed this typo.

**Review by javierdelapuente — DISMISSED** [review]

**Review by mthaddon — DISMISSED** [review]

File: src/haproxy.py [hunk 2]
```diff
@@ -60,14 +66,14 @@ def install(self) -> None:
...
+        """Render the haproxy config and restart the haproxy service.
 
-    def _restart_haproxy_service(self) -> None:
-        """Restart the haporxy service."""
-        systemd.service_restart(HAPROXY_SERVICE)
+        Args:
+            config: charm config
+        """
+        self._render_haproxy_config(config)
+        self._restart_haproxy_service()
```
Comment (gregory-schiano) [inline src/haproxy.py]: Can we use a reload using the `master-worker` option ?
https://docs.haproxy.org/2.8/configuration.html#3.1-master-worker

File: src/haproxy.py [hunk 3]
```diff
@@ -92,9 +98,19 @@ def _render_file(self, path: Path, content: str, mode: int) -> None:
...
+
+        Args:
+            config: charm config
+        """
         with open("templates/haproxy.cfg.j2", "r", encoding="utf-8") as file:
             template = Template(file.read(), keep_trailing_newline=True)
-        rendered = template.render()
+        rendered = template.render(config_global_max_connection=config.global_max_connection)
         self._render_file(HAPROXY_CONFIG, rendered, 0o644)
+        self._restart_haproxy_service()
```
Comment (gregory-schiano) [inline src/haproxy.py]: This will restart twice since another restart is done on L76

**Review by weiiwang01 — DISMISSED** [review]

File: src/state/config.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        Raises:
+            ValueError: When the configured value is not between 0 and "fs.nr_open.
+
+        Returns:
+            int: The validated global_max_connection config.
+        """
+        # No user input so we're disabling bandit rule here as it's considered safe
+        output = check_output(  # nosec: B603
+            ["/usr/sbin/sysctl", "fs.nr_open"], stderr=STDOUT, universal_newlines=True
+        ).splitlines()
```
Comment (gregory-schiano) [inline src/state/config.py]: Catch the `CalledProcessError` and throw a customer or validation exception here

File: src/state/config.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        _, _, fs_nr_open = output[0].partition("=")
+        if global_max_connection < 0 or global_max_connection > int(fs_nr_open.strip()):
```
Comment (gregory-schiano) [inline src/state/config.py]: `fs_nr_open` could be empty string, so add some validation

File: templates/haproxy.cfg.j2 [hunk 6]
```diff
@@ -1,10 +1,9 @@
 global
-    maxconn 4096
+    maxconn {{ config_global_max_connection }}
     user haproxy
     group haproxy
     
-    log /dev/log local0
-    log /dev/log local1 notice
+    log 127.0.0.1:514 local0
```
Comment (gregory-schiano) [inline templates/haproxy.cfg.j2:9]: Should we add a comment to explain why we set this valye ?

File: tests/integration/test_config.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+async def test_config(application: Application, ops_test: OpsTest):
+    """Deploy the charm."""
+    await application.set_config({"global-maxconn": "1024"})
+    await application.model.wait_for_idle(
+        apps=[application.name],
+        idle_period=10,
+        status="active",
+    )
+    _, stdout, _ = await ops_test.juju(
```
Comment (gregory-schiano) [inline tests/integration/test_config.py]: Should just `ops_test.run`

File: src/state/config.py (hunk 4 above)
Comment (Thanhphan1147) [reply src/state/config.py]: Thanks, I added handling for this exception.

File: src/state/config.py (hunk 5 above)
Comment (Thanhphan1147) [reply src/state/config.py]: I realized that `fs.file-max` is what we should use as it's the system-wide hard limit. I also added validation for the output of `.partition`

File: templates/haproxy.cfg.j2 (hunk 6 above)
Comment (Thanhphan1147) [reply templates/haproxy.cfg.j2:9]: Added the comment about rsyslog

File: tests/integration/test_config.py (hunk 7 above)
Comment (Thanhphan1147) [reply tests/integration/test_config.py]: I updated this to use unit.run()

**Review by amandahla — APPROVED** [review]
LGTM

File: src/haproxy.py (hunk 3 above)
Comment (Thanhphan1147) [reply src/haproxy.py]: removed this and updated the rest to `system_reload`

File: src/haproxy.py (hunk 2 above)
Comment (Thanhphan1147) [reply src/haproxy.py]: yep, I switched to using reload

**Review by javierdelapuente — APPROVED** [review]

**Review by weiiwang01 — APPROVED** [review]

File: src/haproxy.py [hunk 8]
```diff
@@ -38,8 +40,8 @@
 logger = logging.getLogger()
 
 
-class HaproxyServiceStartError(Exception):
-    """Error when starting the haproxy service."""
+class HaproxyServiceRestartError(Exception):
+    """Error when restarting the haproxy service."""
```
Comment (gregory-schiano) [inline src/haproxy.py:44]: Seems unused now, but might be needded and renamed to "Reload" instead of restart

File: src/haproxy.py [hunk 9]
```diff
@@ -92,9 +94,17 @@ def _render_file(self, path: Path, content: str, mode: int) -> None:
...
+        """
         with open("templates/haproxy.cfg.j2", "r", encoding="utf-8") as file:
             template = Template(file.read(), keep_trailing_newline=True)
-        rendered = template.render()
+        rendered = template.render(config_global_max_connection=config.global_max_connection)
         self._render_file(HAPROXY_CONFIG, rendered, 0o644)
+
+    def _reload_haproxy_service(self) -> None:
+        """Restart the haporxy service."""
+        systemd.service_reload(HAPROXY_SERVICE)
```
Comment (gregory-schiano) [inline src/haproxy.py:110]: Perhaps do some error handling and use the Reload error here ?
Also the docstring should be changed to "reload" and there's a typo in "haproxy"

File: tests/integration/test_config.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        status="active",
+    )
+
+    action = await application.units[0].run("cat /etc/haproxy/haproxy.cfg", timeout=60)
+    await action.wait()
+
+    code = action.results.get("return-code")
+    stdout = action.results.get("stdout")
+    assert code == 0
+    assert "maxconn 1024" in stdout
```
Comment (gregory-schiano) [inline tests/integration/test_config.py:25]: In the same test you should probably test very high value and invalid value

**Comment by Thanhphan1147** [comment]
I'll merge the PR now and implement the changes requested in a follow-up PR.

**Pull request merged**